File: frogment/src/main/java/ch/grze/frogment/core/parser/Parser.java
package ch.grze.frogment.core.parser;

import android.content.Intent;
import android.os.Bundle;

public class Parser {
    public <T> T getData(String key, T defaultData, Bundle primary, Intent secondary) {
        T data;

        final T primaryData = getDataFrom(key, primary);
        final T secondaryData = getDataFrom(key, secondary);

        data = (primaryData != null) ? primaryData : secondaryData;
        return (data == null) ? defaultData : data;
    }

    public <T> T getData(String key, T defaultData, Bundle primary, Bundle secondary) {
        T state;

        final T primaryData = getDataFrom(key, primary);
        final T secondaryData = getDataFrom(key, secondary);

        state = (primaryData != null) ? primaryData : secondaryData;
        return state == null ? defaultData : state;
    }

    private <T> T getDataFrom(String key, Bundle bundle) {
        if (bundle != null && bundle.getParcelable(key) != null) {
            return (T) bundle.getParcelable(key);
        }

        return null;
    }

    private <T> T getDataFrom(String key, Intent intent) {
        if (intent.getExtras() != null && intent.getExtras().getParcelable(key) != null) {
            return (T) intent.getExtras().getParcelable(key);
        }

        return null;
    }
}

File: frogment/src/main/java/ch/grze/frogment/core/parser/Parser.kt
package ch.grze.frogment.core.parser

import android.content.Intent
import android.os.Bundle
import android.os.Parcelable

class Parser {
    fun <T> getData(key: String, defaultData: T, primary: Bundle?, secondary: Intent): T {
        val primaryData = getDataFrom<T>(key, primary)
        val secondaryData = getDataFrom<T>(key, secondary)

        return primaryData ?: secondaryData ?: defaultData
    }

    fun <T> getData(key: String, defaultData: T, primary: Bundle?, secondary: Bundle?): T {
        val primaryData = getDataFrom<T>(key, primary)
        val secondaryData = getDataFrom<T>(key, secondary)

        return primaryData ?: secondaryData ?: defaultData
    }

    private fun <T> getDataFrom(key: String, bundle: Bundle?): T? {
        return if (bundle?.getParcelable<Parcelable>(key) != null) {
            bundle.getParcelable<Parcelable>(key) as T
        } else null
    }

    private fun <T> getDataFrom(key: String, intent: Intent): T? {
        return if (intent.extras?.getParcelable<Parcelable>(key) != null) {
            intent.extras.getParcelable<Parcelable>(key) as T
        } else null
    }
}
